packages/flutter/lib/src/material/notches.dart

```diff
@@ -0,0 +1,174 @@
+
+    assert(_notchAssertions(host, guest, start, end, fabRadius, notchRadius, notchMargin));
+
+    // If there's no overlap between the guest's margin boundary and the host,
+    // don't make a notch, just return a straight line from start to end.
+    if (!host.overlaps(guest.inflate(notchMargin)))
+      return new Path()..lineTo(end.dx, end.dy);
+
+    // We build a path for the notch from 3 segments:
+    // Segment A - a Bezier curve from the host's top edge to segment B.
+    // Segment B - an arc with radius notchRadius.
+    // Segment C - a Bezier curver from segment B back to the host's top edge.
+    //
+    // A detailed explanation and the derivation of the formulas below is
+    // available at: https://goo.gl/Ufzrqn
+
+    const double s1 = 15.0;
+    const double s2 = 1.0;
+
+    final double r = notchRadius;
+    final double a = -1.0 * r - s2;
+    final double b = host.top - guest.center.dy;
+
+    final double n2 = math.sqrt(b * b * r * r * (a * a + b * b - r * r));
+    final double p2xA = ((a * r * r) - n2) / (a * a + b * b);
+    final double p2xB = ((a * r * r) + n2) / (a * a + b * b);
+    final double p2yA = math.sqrt(r * r - p2xA * p2xA);
+    final double p2yB = math.sqrt(r * r - p2xB * p2xB);
+
+    final List<Offset> p = new List<Offset>(6);
+
+    // p0, p1, and p2 are the control points for segment A.
+    p[0] = new Offset(a - s1, b);
+    p[1] = new Offset(a, b);
+    final double cmp = b < 0 ? -1.0 : 1.0;
+    p[2] = cmp * p2yA > cmp * p2yB ? new Offset(p2xA, p2yA) : new Offset(p2xB, p2yB);
+
+    // p3, p4, and p5 are the control points for segment B, which is a mirror
+    // of segment A around the y axis.
+    p[3] = new Offset(-1.0 * p[2].dx, p[2].dy);
+    p[4] = new Offset(-1.0 * p[1].dx, p[1].dy);
+    p[5] = new Offset(-1.0 * p[0].dx, p[0].dy);
+
+    // translate all points back to the absolute coordinate system.
+    for (int i = 0; i < p.length; i += 1)
+      p[i] += guest.center;
+
+    return new Path()
+      ..lineTo(p[0].dx, p[0].dy)
+      ..quadraticBezierTo(p[1].dx, p[1].dy, p[2].dx, p[2].dy)
+      ..arcToPoint(
+        p[3],
+        radius: new Radius.circular(notchRadius),
+        clockwise: false,
+      )
+      ..quadraticBezierTo(p[4].dx, p[4].dy, p[5].dx, p[5].dy)
+      ..lineTo(end.dx, end.dy);
+  }
+
+  bool _notchAssertions(Rect host, Rect guest, Offset start, Offset end,
+    double fabRadius, double notchRadius, double notchMargin) {
+    if (end.dy != host.top)
+      throw new FlutterError(
+        'The notch of the guest must end at the top edge of the host.\n'
+        'The notch\'s path end point: $end is not in the top edge of $host'
+      );
+
+    if (start.dy != host.top)
+      throw new FlutterError(
+        'The notch of the guest must start at the top edge of the host.\n'
+        'The notch\'s path start point: $start is not in the top edge of $host'
+      );
+
+    if (guest.center.dx - notchRadius < start.dx)
```

Comment:
Why not just have CircularNotch swap start and end if necessary?

Reply:
done